homeassistant/components/bryant_evolution/coordinator.py

```diff
@@ -0,0 +1,190 @@
+
+    # Note: directly accessing the data member of this class is discouraged.
+    # Prefer using the provided read_ functions, which are less error-prone.
+
+    def __init__(
+        self, hass: HomeAssistant, tty: str, system_zones: list[tuple[int, int]]
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            logger=_LOGGER,
+            name=f"Bryant Evolution at {tty}",
+            update_interval=timedelta(minutes=1),
+            update_method=lambda: _read_evolution_state(
+                self.system_zone_tuples, self.tty
+            ),
+            always_update=False,
+        )
+        self.system_zone_tuples = system_zones
+        self.tty = tty
+
+    def read_current_temperature(self, sys_id: int, zone_id: int) -> int:
+        """Return the current temperature."""
+        return self.data.systems[sys_id].zones[zone_id].curr_temp
+
+    def read_fan_mode(self, sys_id: int) -> str:
+        """Return the current fan mode."""
+        return self.data.systems[sys_id].fan_mode
+
+    def read_hvac_mode(self, sys_id: int) -> HVACMode:
+        """Return the current HVAC mode."""
+        evolution_mode = self.data.systems[sys_id].hvac_mode
+        mode_enum = {
+            "HEAT": HVACMode.HEAT,
+            "COOL": HVACMode.COOL,
+            "AUTO": HVACMode.HEAT_COOL,
+            "OFF": HVACMode.OFF,
+        }.get(evolution_mode.upper())
+        if mode_enum is None:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="failed_to_parse_hvac_mode",
+                translation_placeholders={"mode": evolution_mode},
+            )
+        return mode_enum
+
+    def read_target_temperatures(
+        self, sys_id: int, zone_id: int
+    ) -> tuple[int | None, int | None, int | None]:
+        """Return the target temperatures.
+
+        Specifically, returns a tuple with three integers. It will be either:
+        1. (target-temp, None, None) if the mode is HEAT or COOL
+        2. (None, heating-setpoint, cooling-setpoint) if the mode is HEAT_COOL.
+        """
+        mode = self.read_hvac_mode(sys_id)
+        # Return target_temperature or target_temperature_{high, low} based on mode.
+        match mode:
+            case HVACMode.HEAT:
+                return (self.data.systems[sys_id].zones[zone_id].htsp, None, None)
+            case HVACMode.COOL:
+                return (self.data.systems[sys_id].zones[zone_id].clsp, None, None)
+            case HVACMode.HEAT_COOL:
+                htsp = self.data.systems[sys_id].zones[zone_id].htsp
+                clsp = self.data.systems[sys_id].zones[zone_id].clsp
+                return (None, htsp, clsp)
+            case HVACMode.OFF:
+                return (None, None, None)
+            case _:
+                raise HomeAssistantError(f"Unknown HVAC mode {mode}")
```

Comment:
You're right, but:

mypy can't figure that out ("homeassistant/components/bryant_evolution/coordinator.py:141: error: Missing return statement [return]" without this case)

I think it's good to avoid this kind of cross-function reasoning where possible. https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ is a great article.

Reply:
I don't really like that this function returns a tuple. You don't need this complexity, create properties in the climate entity. Something like this:
```
@property
def target_temperature(self) -> HVACMode | None

if self.hvac_mode is HVACMode.HEAT:
    return self.coordinator.data.systems[sys_id].zones[zone_id].htsp
if self.hvac_mode is HVACMode.COOL:
    return self.coordinator.data.systems[sys_id].zones[zone_id].clsp

@property
def target_temperature_low(self) -> float | None
if self.hvac_mode is HVACMode.HEAT_COOL
    return self.coordinator.data.systems[sys_id].zones[zone_id].htsp

@property
def target_temperature_high(self) -> float | None
if self.hvac_mode is HVACMode.HEAT_COOL
    return self.coordinator.data.systems[sys_id].zones[zone_id].clsp

```
 I think that looks a loot cleaner and each property only accounts for the cases it has to and nothing else

Reply:
I think this depends on the discussion of whether we can use properties, above.